source/whats-new.txt

```diff
@@ -0,0 +1,123 @@
+.. _cpp-whats-new:
+
+==========
+What's New
+==========
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 1
+   :class: singlecol
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: update, new feature, deprecation, upgrade, driver
+
+Learn what's new in:
+
+* :ref:`Version 3.10 <version-3.10>`
+* :ref:`Version 3.9 <version-3.9>`
+
+.. _upcoming-breaking-changes:
+
+.. _version-3.10:
+
+What's New in 3.10
+------------------
+
+.. warning:: Breaking Changes
+
+   This driver version contains the following breaking changes:
+   
+   - Drops support for ``find_package(libbsoncxx)`` and ``find_package(libmongocxx)``.
+     Use ``find_package(bsoncxx)`` and ``find_package(mongocxx)``
+     instead.
+   - Drops support for the ``LIBBSONCXX_*`` and ``LIBMONGOCXX_*`` CMake variables
+     provided by the legacy CMake package config files. Use the ``mongo::bsoncxx_*`` and
+     ``mongo::mongocxx_*`` CMake targets instead.
+   - Removes the experimental C++ standard library (``BSONCXX_POLY_USE_STD_EXPERIMENTAL``)
+     as a polyfill option.
+
+The v3.10.0 release includes the following new features:
+
+- Adds forward headers that provide non-defining declarations of ``bsoncxx`` and
+  ``mongocxx`` class types. The driver does not support user-defined forward declarations
+  of any library entity. To obtain the declaration or definition of a library entity,
+  always include the corresponding header.
+- The CMake option ``ENABLE_BSONCXX_USE_POLY_IMPLS`` allows selecting
+  ``bsoncxx`` implementations of C++17 polyfills by default when no polyfill library is
+  requested. This option is ``OFF`` by default.
+- The CMake option ``BSONCXX_POLY_USE_IMPLS`` allows selecting ``bsoncxx`` implementations
+  of C++17 polyfills instead of external libraries or the C++ standard library. This
+  option is ``OFF`` by default.
+- The ``bsoncxx::v_noabi`` and ``mongocxx::v_noabi`` namespaces are no longer declared inline.
+  Root namespace declarations are still supported and expected to remain equivalent to
+  their prior definitions. For example, ``bsoncxx::document`` is still equivalent to
+  ``bsoncxx::v_noabi::document``, and ``mongocxx::client`` is still equivalent to
+  ``mongocxx::v_noabi::client``. Argument-dependent lookup and template instantiations
+  are expected to remain equivalent to their prior behavior.
```

Comment:
Good question! This language comes directly from the [release notes](https://github.com/mongodb/mongo-cxx-driver/releases/tag/r3.10.0), which makes me think this is as much as they know? I can bring it up with the tech reviewer

Reply:
This phrasing was me being honest regarding my level of confidence concerning the backward compatibility of this refactor.

C++ argument-dependent lookup and template instantiation rules are infamously complicated; despite my best efforts, even if I am confident that the _normal_ and _expected_ usage of the C++ Driver interfaces will remain compatible, I couldn't be 100% confident that absolutely _no_ user code won't be broken due to some unforeseen (likely esoteric) usage pattern (see: Hyrum's Law). This concern is why I also included in the release notes:

> If this change does inadvertently break source or binary compatibility, please submit a bug report.

The intent and expectation is that this does not break user code. I do not believe anyone has submiited such a bug report yet since the release. If we'd prefer this language be more definitive in the "What's New" documentation, that is fine by me.

Reply:
Swapped in the bit about submitting a bug report. I considered adding it as as general recommendation at the top of the page, but didn't think was necessary, and thought it could even be more confusing given all of the actual breaking changes on the page.